lib/crucible/array.rs

```diff
@@ -0,0 +1,55 @@
+use core::marker::PhantomData;
+use crate::symbolic::Symbolic;
+
+#[derive(Copy)]
+pub struct Array<T>(PhantomData<T>);
+
+// NB: `T: Copy`, not `T: Clone`.  Using `clone` would require us to know all populated indices of
+// the array, which we don't.
```

Comment:
`T: Clone` would be incorrect in general, though it's likely that there's no way to expose the difference at the moment (and there may not ever be).

Since `Array`s are unbounded (more or less), we can't clone the collection by cloning each element, like most container types do.  The only thing we can do is copy the whole `Array`, which is effectively the same as copying each of its elements.  But copying non-`Copy` values is unsafe because it can violate global invariants, such as "each `Box<T>` points to a distinct heap allocation" or "the refcount of each `Rc<T>` is equal to the number of aliases in existence".  So we restrict `Array`'s `Clone` impl to `T: Copy` because we can't implement it properly given only `T: Clone`.

I'll update the comment to be a bit more descriptive.